PINCache/PINCache.m

```diff
@@ -152,157 +141,85 @@ - (void)setObject:(id <NSCoding>)object forKey:(NSString *)key block:(PINCacheOb
 {
     if (!key || !object)
         return;
+  
+    PINOperationGroup *group = [PINOperationGroup asyncOperationGroupWithQueue:self.operationQueue];
     
-    dispatch_group_t group = nil;
-    PINMemoryCacheObjectBlock memBlock = nil;
-    PINDiskCacheObjectBlock diskBlock = nil;
-    
-    if (block) {
-        group = dispatch_group_create();
-        dispatch_group_enter(group);
-        dispatch_group_enter(group);
-        
-        memBlock = ^(PINMemoryCache *memoryCache, NSString *memoryCacheKey, id memoryCacheObject) {
-            dispatch_group_leave(group);
-        };
-        
-        diskBlock = ^(PINDiskCache *diskCache, NSString *diskCacheKey, id <NSCoding> memoryCacheObject) {
-            dispatch_group_leave(group);
-        };
-    }
-    
-    [_memoryCache setObject:object forKey:key block:memBlock];
-    [_diskCache setObject:object forKey:key block:diskBlock];
+    [group addOperation:^{
+        [_memoryCache setObject:object forKey:key];
+    }];
+    [group addOperation:^{
+        [_diskCache setObject:object forKey:key];
+    }];
+    __weak PINCache *weakSelf = self;
+    [group setCompletion:^{
+        PINCache *strongSelf = weakSelf;
+        block(strongSelf, key, object);
+    }];
     
-    if (group) {
-        __weak PINCache *weakSelf = self;
-        dispatch_group_notify(group, _concurrentQueue, ^{
-            PINCache *strongSelf = weakSelf;
-            if (strongSelf)
-                block(strongSelf, key, object);
-        });
-        
-#if !OS_OBJECT_USE_OBJC
-        dispatch_release(group);
-#endif
-    }
+    [group start];
 }
 
 - (void)removeObjectForKey:(NSString *)key block:(PINCacheObjectBlock)block
 {
     if (!key)
         return;
     
-    dispatch_group_t group = nil;
-    PINMemoryCacheObjectBlock memBlock = nil;
-    PINDiskCacheObjectBlock diskBlock = nil;
-    
-    if (block) {
-        group = dispatch_group_create();
-        dispatch_group_enter(group);
-        dispatch_group_enter(group);
-        
-        memBlock = ^(PINMemoryCache *memoryCache, NSString *memoryCacheKey, id memoryCacheObject) {
-            dispatch_group_leave(group);
-        };
-        
-        diskBlock = ^(PINDiskCache *diskCache, NSString *diskCacheKey, id <NSCoding> memoryCacheObject) {
-            dispatch_group_leave(group);
-        };
-    }
+    PINOperationGroup *group = [PINOperationGroup asyncOperationGroupWithQueue:self.operationQueue];
     
-    [_memoryCache removeObjectForKey:key block:memBlock];
-    [_diskCache removeObjectForKey:key block:diskBlock];
+    [group addOperation:^{
+        [_memoryCache removeObjectForKey:key];
```

Comment:
Is it bad that this captures self by using instance variable?  This seems maybe OK because the operations should have finite lifecycle.  If it is OK, we should remove the weak / strongSelf below in the completion, since it wouldn't be worth using a weak variable if everything but the completion block will keep self alive.